### ISSUE TITLE
Pending Reads not completing

### ISSUE THREAD

**Issue by ThomasSchuh** [microsoft/FASTER#378]
Hi there,

I'm having the issue that Read() on keys that have been written to disk give me Status.Pending when i try to read them again. Tested a bunch already and had a look at the implementations in your unit tests and benchmark project but i can't see what do wrong. I actually tried to set up the benchmark project similar to mine and ran into the same issue.

Tried closing the session after the writes and crating a new one before reading, as well as session.CompletePending() and 
 session.Refresh() but nothing appear to impact my case at all. Also saw context.FinalizeRead() somewhere in the tests so i set that up with the same result.

In the serializers Deserialize() call i seem to get an empty stream for those items, not sure why that is (and if it happens consistently. Didn't follow up on that detail).

I pushed my test project [here](https://github.com/ThomasSchuh/FasterTest). It will Upsert a bunch of items and then tries to read them. From what i saw i'd say it can consistently read what is still in memory but all other items are stuck on pending.

Would be awesome if you could have a quick look and point me in the right direction. 

Regards,
Thomas

**Comment by badrishc** [comment]
The way FASTER returns pending results, when using the sync API, is via completion callbacks. These are provided via `IFunctions` argument to `NewSession`. You are using the default `SimpleFunctions<CacheKey, CacheValue, StoreContext>` which do nothing in the callback. You need to override this class and provide your own callback as follows:

```cs
public class StoreFunctions : SimpleFunctions<CacheKey, CacheValue, StoreContext>
{
   public override void ReadCompletionCallback(ref CacheKey key, ref CacheValue input, ref CacheValue output, StoreContext ctx, Status status)
      => ctx.Populate(ref status, ref output);
}
``` 

In the above callback, we simply populate the context with the status and output.

You can then instantiate the session as follows:

```cs
var session = _fasterStore.NewSession(new StoreFunctions());
```

**Comment by badrishc** [comment]
The other option would be to use the async Read API: `ReadAsync`. See example [here](https://github.com/microsoft/FASTER/blob/master/cs/samples/StoreAsyncApi/Program.cs).

**Comment by ThomasSchuh** [comment]
Thanks a lot for the quick response. Really appreciate it.

I could solve the "Pending" issue with the ReadCompletionCallback. But that led me back to my inital problem which is that the Serializer at some point gets an empty stream. I played around with that the whole day and from my observations I'd say I get the empty stream once I read a Page worth of data from disk. Unfortunately I don't really understand what impact the Page size should have so I'm kind of guessing here.

I updated the example repo. With that setup i quickly run into the empty stream issue a few times followed by a way to big stream. Thought it might be just to small pages but making them bigger only delays the issue. At some point I always get empty streams.
